src/allocator.rs

```diff
@@ -245,6 +256,44 @@ impl Allocator {
         number_from_u8(self.atom(node))
     }
 
+    pub fn g1(&self, node: NodePtr) -> Result<G1Projective, EvalErr> {
```

Comment:
If we can get this change in before the hard fork, we can be a lot more accurate by charging for serialization/deserialization. Converting `g1` to/from `bytes` is very expensive, so a DOS attack would be to convert back and forth from `g1` to `bytes`. A genuine usage would just do a bunch of `g1` calculations with no intermediate conversions. It would be fantastic to be able to charge less for this. But it's obviously a hard fork.

All this same reasoning could be done with `Number`.

Reply:
right. I'm a bit hesitant to expose so much implementation details in the consensus logic (i.e. cost model). You're right that there's a risk of not doing that. I'm thinking that one possibility would be to convert lazily, but also cache all representations of a value, so at least you can't convert the same value back-and-forth

Reply:
> You're right that there's a risk of not doing that.

It's not so much a risk (we'll still charge as IF the conversion is done every time) as a lack of reward for not trying to DOS. If we don't do this, then users doing legit computation that doesn't require conversion still have to pay the price they would if they were trying to DOS even though we have the capability to better notice this. It's just leaving potential agony cost on the table. But it's not the end of the world. It just makes actual DOS attempts more noticeable.